specs/lemmas/Spec.Hash.Incremental.fst

```diff
@@ -596,10 +596,291 @@ let md_is_hash_incremental
      }
 #pop-options
 
-let hash_is_hash_incremental (a: hash_alg) (input: bytes { S.length input <= max_input_length a }):
+let sha3_is_incremental1
+  (a: sha3_alg)
+  (input: bytes): Lemma (hash_incremental a input `S.equal` (
+    let s = Lib.Sequence.create 25 (u64 0), () in
+    let rateInBytes = 1088 / 8 in
+    let delimitedSuffix = byte 0x06 in
+    let bs, l = UpdateMulti.split_at_last rateInBytes input in
+    let s, _ = update_multi a s bs in
+    let s = Spec.SHA3.absorb_last delimitedSuffix rateInBytes (S.length l) l s, () in
+    finish a s))
+=
+  calc (S.equal) {
+    hash_incremental a input;
+  (S.equal) { } (
+    let s = Lib.Sequence.create 25 (u64 0), () in
+    let bs, l = split_blocks a input in
+    let s = update_multi a s bs in
+    let s = update_last a s (S.length bs) l in
+    finish a s
+  );
+  (S.equal) { } (
+    let s = Lib.Sequence.create 25 (u64 0), () in
```

Comment:
do you really need to duplicate so much code for each `calc` step?
```
 let s = Lib.Sequence.create 25 (u64 0), () in
 let rateInBytes = 1088 / 8 in
 let delimitedSuffix = byte 0x06 in
```
can these proofs be used for other instances of sha3? if so, should we make them generic?

Reply:
> do you really need to duplicate so much code for each calc step?
it took me a lot of steps to get to the point where I understood what the problem was (the lack of type equality between `words_state SHA3_256` and `Spec.SHA3.state` so I think I broke it down into *way* too many steps!

> if so, should we make them generic?
I structured things in such a way that I can easily make them generic... but in a followup PR :-)